### PR TITLE
345 add see more truncation pattern for typography

### PR DESCRIPTION
**Pull request by gd2910**
## Summary of the changes
Added a truncation pattern for the 'body' variant of ic-typogrpahy. When max-lines property is set, the user will be able to click a 'see more...' link that will display the full text, and click the 'see loess...' link to truncate the text once more.

## Related issue
#345 

## Checklist
- [x] I have added relevant unit and visual regression tests.
- [x] I have manually accessibility tested any changes, if relevant.
- [x] I have ensured any changes match the Figma component library.

### CODE REVIEW

File: packages/react/src/stories/ic-typography.stories.mdx [hunk 1]
```diff
@@ -42,28 +47,69 @@ import readme from "../../../web-components/src/components/ic-typography/readme.
...
-    <IcTypography applyVerticalMargins variant="h3">Medium title</IcTypography>
-    <IcTypography applyVerticalMargins variant="h4">Small title</IcTypography>
-    <IcTypography applyVerticalMargins variant="subtitle-large">Subtitle 1</IcTypography>
-    <IcTypography applyVerticalMargins variant="subtitle-small">Subtitle 2</IcTypography>
-    <IcTypography applyVerticalMargins variant="body">Body text</IcTypography>
-    <IcTypography applyVerticalMargins variant="caption">Caption text</IcTypography>
-    <IcTypography applyVerticalMargins variant="caption-uppercase">Caption uppercase</IcTypography>
-    <IcTypography applyVerticalMargins variant="code-large">Code large</IcTypography>
-    <IcTypography applyVerticalMargins variant="code-small">Code small</IcTypography>
-    <IcTypography applyVerticalMargins variant="code-extra-small">Code extra small</IcTypography>
```
Comment (GCHQ-Developer-112) [inline packages/react/src/stories/ic-typography.stories.mdx:64]: The code variants have been removed from this story. Was this intended?

File: packages/web-components/src/components/ic-typography/ic-typography.spec.ts [hunk 2]
```diff
@@ -291,4 +291,35 @@ describe("ic-typography component", () => {
...
+    });
+
+    const link = await document
+      .querySelector("ic-typography")
+      .shadowRoot.querySelector("ic-link");
+
+    link.click();
+
+    page.waitForChanges();
+
```
Comment (ad9242) [inline packages/web-components/src/components/ic-typography/ic-typography.spec.ts:307]: should we add a check here that it has "See less" visible?

File: packages/web-components/src/components/ic-typography/ic-typography.tsx [hunk 3]
```diff
@@ -20,8 +20,66 @@ export class Typography {
...
+    this.maxLines > 0 && this.variant == "body"
+      ? (this.truncStyle = {
+          "max-height": this.boxHeight(this.maxLines),
+          overflow: "hidden",
+        })
+      : null;
+  }
+
+  private boxHeight = (lines: number): any => {
+    return (lines * 1.5).toString() + "rem";
```
Comment (ad9242) [inline packages/web-components/src/components/ic-typography/ic-typography.tsx:42]: where does the 1.5 come from? should we have a global variable for this

**Review by ad9242 — COMMENTED** [review]
looks good - just a couple of comments

File: packages/web-components/src/components/ic-typography/ic-typography.tsx (hunk 3 above)
Comment (gd2910) [reply packages/web-components/src/components/ic-typography/ic-typography.tsx:42]: the 1.5 is the global line-height of the body variant to calculate the correct size box, using the design token wouldn't work in this calc I don't think?

**Comment by GCHQ-Developer-530** [comment]
Not sure if this is intended or not, but when see more is clicked the link styling that usually just appears on hover, appears all the time until I click away from it.

https://user-images.githubusercontent.com/111882034/234849419-2fb41433-906d-42aa-9558-6fd5f6cb21f5.mov

**Comment by gd2910** [comment]
> Not sure if this is intended or not, but when see more is clicked the link styling that usually just appears on hover, appears all the time until I click away from it.
> 
>  Screen.Recording.2023-04-27.at.12.30.36.mov

Initially unintended, and had the same issue whilst using ic-link, but after investigation I think it should be the intended use,  because the focus stays on the button and keyboard users can easily click the button again. Waiting to hear from design whether this is acceptable, or may have to force focus off the button on click.

File: packages/react/src/stories/ic-typography.stories.mdx [hunk 4]
```diff
@@ -39,25 +47,69 @@ import readme from "../../../web-components/src/components/ic-typography/readme.
...
   </Story>
 </Canvas>
 
+### Truncation
 
-
+<Canvas>
+  <Story name="Truncation">
+    <IcTypography max-lines="1" variant="body">
+      Body of text that is truncated to one line. Click the 'See more...' link
```
Comment (ad9242) [inline packages/react/src/stories/ic-typography.stories.mdx:102]:
```suggestion
      Body of text that is truncated to one line. Click the 'See more' link
```

File: packages/react/src/stories/ic-typography.stories.mdx [hunk 5]
```diff
@@ -39,25 +47,69 @@ import readme from "../../../web-components/src/components/ic-typography/readme.
...
+      to expand the text, then click 'See less...' to truncate the text once
```
Comment (ad9242) [inline packages/react/src/stories/ic-typography.stories.mdx:103]:
```suggestion
      to expand the text, then click 'See less' to truncate the text once
```

File: packages/web-components/src/components/ic-typography/ic-typography.stories.mdx [hunk 6]
```diff
@@ -84,6 +85,37 @@ import readme from "./readme.md";
...
+  </Story>
+</Canvas>
+
+### Truncation
+
+<Canvas>
+  <Story name="Truncation">
+    {html`
+      <ic-typography max-lines="1" variant="body">
+        Body of text that is truncated to one line. Click the 'See more...' link
```
Comment (ad9242) [inline packages/web-components/src/components/ic-typography/ic-typography.stories.mdx:107]:
```suggestion
        Body of text that is truncated to one line. Click the 'See more' link
```

File: packages/web-components/src/components/ic-typography/ic-typography.stories.mdx [hunk 7]
```diff
@@ -84,6 +85,37 @@ import readme from "./readme.md";
...
+        to expand the text, then click 'See less...' to truncate the text once
```
Comment (ad9242) [inline packages/web-components/src/components/ic-typography/ic-typography.stories.mdx:108]:
```suggestion
        to expand the text, then click 'See less' to truncate the text once
```

**Review by ad9242 — CHANGES_REQUESTED** [review]
A few little comments, plus I think a rebase is required

**Comment by ad9242** [comment]
My only other comment would be how it should behave if the text fits ok, as currently it would still have a redundant "See more" eg 

image

maybe @DBD324 can advise?

**Review by ad9242 — DISMISSED** [review]

File: packages/web-components/src/components/ic-typography/ic-typography.stories.mdx [hunk 8]
```diff
@@ -23,7 +23,7 @@ import readme from "./readme.md";
 ### Variants
 
 <Canvas>
-  <Story name="Variants">
+  <Story name="Variants" parameters={{ loki: { skip: true } }}>
```
Comment (GCHQ-Developer-530) [inline packages/web-components/src/components/ic-typography/ic-typography.stories.mdx:26]: The loki skip on this and vertical margins can be removed now that a fix was merged for the visual tests failing 😊

**Review by GCHQ-Developer-530 — DISMISSED** [review]

**Comment by jd239** [comment]
> Not sure if this is intended or not, but when see more is clicked the link styling that usually just appears on hover, appears all the time until I click away from it.
> 
>  Screen.Recording.2023-04-27.at.12.30.36.mov

I'm still getting the behaviour. Expected hover state is still present. Could it be because the button is still focused after the click and the focus state is the same as the hover state?

**Comment by jd239** [comment]
The last AC suggests there should be no ellipsis if the extra text starts in a new paragraph but is not cut off from the overflow. Not sure how we'll do this to be honest.

AC

> Given that the See more/See less component includes flexibility in its logic
> When there is additional content - but on a separate paragraph
> And the last visible line is short enough the house all the content
> Then no ellipses is shown, and the ‘See more’ component appears ~~after one space-bar key width from the final full stop (see > attached photo: Example 2)~~

Wonder if it's worth design knowing about this..

File: packages/web-components/src/components/ic-typography/ic-typography.css [hunk 9]
```diff
@@ -125,3 +126,47 @@
...
+  text-decoration: underline;
+  font-weight: var(--ic-font-weight-bold);
+  transition: var(--ic-easing-transition-fast);
+}
+
+.trunc-btn:hover,
+.trunc-btn:focus {
+  outline: none;
+  border-bottom: 0.25rem solid !important;
+  margin-bottom: -0.25rem !important;
```
Comment (jd239) [inline packages/web-components/src/components/ic-typography/ic-typography.css:158]: Can use spacing tokens here. `--ic-space-xxs`

File: packages/web-components/src/components/ic-typography/ic-typography.e2e.ts [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    await page.waitForChanges();
+
+    await page.setViewport({
+      width: 1600,
+      height: 480,
+    });
+
+    await page.waitForChanges();
+    // page.debugger()
+    // await page.evaluate( () => { debugger; } );
```
Comment (jd239) [inline packages/web-components/src/components/ic-typography/ic-typography.e2e.ts:44]: Remove commented out code

File: packages/web-components/src/components/ic-typography/ic-typography.tsx [hunk 11]
```diff
@@ -20,8 +21,107 @@ export class Typography {
...
+  private truncatedHeight: number = 0;
+  private marker: HTMLElement;
+
+  @Watch("expanded")
+  watchExpandedHandler(): void {
+    if (this.expanded) {
+      this.el.setAttribute("style", `--truncation-max-lines: initial`);
+    } else {
+      this.el.setAttribute("style", `--truncation-max-lines: ${this.maxLines}`);
+    }
```
Comment (jd239) [inline packages/web-components/src/components/ic-typography/ic-typography.tsx:47]: Could probably put this on one line
```js
this.el.setAttribute("style", `--truncation-max-lines: ${this.expanded ? 'initial' : this.maxLines}`)
```

File: packages/web-components/src/components/ic-typography/ic-typography.tsx [hunk 12]
```diff
@@ -20,8 +21,107 @@ export class Typography {
...
+  }
+
+  private toggleExpanded = () => {
+    this.expanded = !this.expanded;
+  };
+
+  private checkMaxLines = (height: number) => {
+    const numLines = Math.floor(height / 24);
```
Comment (jd239) [inline packages/web-components/src/components/ic-typography/ic-typography.tsx:55]: Could we add a comment where 24 comes from?

**Review by jd239 — CHANGES_REQUESTED** [review]
Just a couple of comments

**Review by jd239 — CHANGES_REQUESTED** [review]
Just a couple of comments

**Comment by ad9242** [comment]
> > Not sure if this is intended or not, but when see more is clicked the link styling that usually just appears on hover, appears all the time until I click away from it.
> > Screen.Recording.2023-04-27.at.12.30.36.mov
> 
> I'm still getting the behaviour. Expected hover state is still present. Could it be because the button is still focused after the click and the focus state is the same as the hover state?

yes, the button still has the focus, so will get the focus styling applied. not sure how to resolve as otherwise the focus state would not be applied when you tab to it

**Comment by ad9242** [comment]
> The last AC suggests there should be no ellipsis if the extra text starts in a new paragraph but text is not cut off from the overflow. Not sure how we'll do this to be honest.
> 
> AC:
> 
> > Given that the See more/See less component includes flexibility in its logic
> > When there is additional content - but on a separate paragraph
> > And the last visible line is short enough the house all the content
> > Then no ellipses is shown, and the ‘See more’ component appears ~after one space-bar key width from the final full stop (see > attached photo: Example 2)~
> 
> Capture
> 
> Wonder if it's worth design knowing about this..

design approved

**Review by GCHQ-Developer-530 — APPROVED** [review]

**Review by MI6-451 — APPROVED** [review]

**Review by jd239 — APPROVED** [review]

**Pull request merged**